### PR TITLE
Issue 4524 - Project Images Endpoints

### LINKED ISSUE #4524: Allow projects to have images
### Description
We should allow projects to have images as described in [here](https://github.com/3drepo/3D-Repo-Product-Team/issues/292). 

Changes:

####  Get a project image:

GET v5/teamspaces/{teamspace}/projects/{project}/image

response: binary stream of the image
permission: anyone who has access to the project

#### Update a project image:

PUT v5/teamspaces/{teamspace}/projects/{project}/image

payload: multi-part form with the key palue pair: "file" : file in binary form
response: OK for success or an error code if it fails
permission: Project admin (or teamspace admin)

image

#### Delete a project image:

DELETE v5/teamspaces/{teamspace}/projects/{project}/image

response: OK for success or an error code if it fails
permission: Project admin (or teamspace admin)

### Goals
- [ ] User should be able to add a project image.
- [ ] User should be able to edit a project image.
- [ ] User should be able to delete a project image.
- [ ] User should be able to retrieve project image.

### Tasks
GET endpoint
- [ ] Should retrieve the image of the project and return a binary stream
- [ ] Should return fileNotFound if there is no image

PUT endpoint
- [ ] Should accept the binary stream of the image and validate the data
- [ ] Should store the image in fileshare as well as the ref projects.ref
- [ ] If there is an existing image it should be removed from fileshare as well as its ref should be removed from projects.ref

DELETE endpoint
- [ ] Should remove the image from filestream
- [ ] Should remove the ref from projects.ref

### PR DESCRIPTION
**Pull request by ChristosTsiotsias**
This fixes #4524 

#### Description
Endpoints for fetching, updating and deleting project images.
Swagger documentation, unit and e2e tests added.

#### Test cases
GET endpoint should retrieve the project image and return it in binary form
PUT endpoint should validate the payload (binary image), store the image in the filestream and insert the ref in the db. Also remove any previous refs and images from the fileshare
DELETE endpoint should remove an existing image from the fileshare and remove the ref from the db

### CODE REVIEW

File: backend/tests/v5/unit/processors/teamspaces/projects/projects.test.js [hunk 1]
```diff
@@ -157,10 +164,93 @@ const testGetAllTemplates = () => {
 	});
 };
 
+const testGetImage = () => {
+	describe('Get image', () => {
+		test('should get image stream', async () => {
```
Comment (carmenfan) [inline backend/tests/v5/unit/processors/teamspaces/projects/projects.test.js:169]: from your implementation we're not gettig image stream, we're getting the blob in this entirety?

**Review by carmenfan — CHANGES_REQUESTED** [review]

File: backend/src/v5/middleware/dataConverter/inputs/teamspaces/projects/index.js [hunk 2]
```diff
@@ -44,4 +46,27 @@ Projects.validateProjectData = async (req, res, next) => {
...
+
+		await next();
+	} catch (err) {
+		respond(req, res, err);
+	}
+};
+
+const validateProjectImage = async (req, res, next) => {
+	try {
+		if (!req.file) throw createResponseCode(templates.invalidArguments, 'A file must be provided');
```
[the remark this answers is not in the document]
Comment (ChristosTsiotsias) [reply backend/src/v5/middleware/dataConverter/inputs/teamspaces/projects/index.js:62]: it seems to pass singleFileMulterPromise without an exception being thrown

File: backend/src/v5/routes/teamspaces/projects/projects.js [hunk 3]
```diff
@@ -286,6 +325,107 @@ const establishRoutes = () => {
...
+	 *         in: path
+	 *         required: true
+	 *         schema:
+	 *           type: string
+	 *           format: uuid
+	 *     responses:
+	 *       401:
+	 *         $ref: "#/components/responses/notLoggedIn"
+	 *       200:
+	 *         description: Deletes the image of the project
```
[the remark this answers is not in the document]
Comment (ChristosTsiotsias) [reply backend/src/v5/routes/teamspaces/projects/projects.js:425]: do you mean like this ? 

image

I dont think I've seen it in another endpoint

[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v5/routes/teamspaces/projects/projects.js:425]: yea

File: backend/src/v5/middleware/dataConverter/inputs/teamspaces/projects/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v5/middleware/dataConverter/inputs/teamspaces/projects/index.js:62]: in that case we either:

1. we check in `singleFileMulterPromise`  to ensure file is there, or;
2. add a check in `singleFileUpload` and throw an error there.


Either way I don't think we need another middleware that is a duplicate in 2 places to check the file needs to exist

File: backend/src/v5/routes/teamspaces/projects/projects.js [hunk 4]
```diff
@@ -88,6 +90,44 @@ const getProject = async (req, res) => {
...
+		const fileExt = await fileExtensionFromBuffer(buffer);
+		req.params.format = fileExt || 'png';
+		respond(req, res, templates.ok, buffer);
+	} catch (err) {
+		// istanbul ignore next
+		respond(req, res, err);
+	}
+};
+
+const uploadImage = async (req, res) => {
```
Comment (carmenfan) [inline backend/src/v5/routes/teamspaces/projects/projects.js:107]: upload isn't really the action here anymore, it's more like `updateImage`

**Review by carmenfan — CHANGES_REQUESTED** [review]

File: backend/src/v5/middleware/dataConverter/multer.js [hunk 5]
```diff
@@ -60,13 +60,11 @@ const imageFilter = (req, file, cb) => {
 
 const ensureFileIsImage = async (req, res, next) => {
 	const fileBuffer = req.file?.buffer;
-	if (fileBuffer) {
-		const ext = await fileExtensionFromBuffer(fileBuffer);
+	const ext = await fileExtensionFromBuffer(fileBuffer);
```
[the remark this answers is not in the document]
Comment (ChristosTsiotsias) [reply backend/src/v5/middleware/dataConverter/multer.js:63]: if filebuffer is undefined then ext will be undefined and will throw unsupportedFileFormat error

**Review by carmenfan — APPROVED** [review]

**Pull request merged**